library/src/permutation/permutation_cpu_reference_impl.hpp

```diff
@@ -23,80 +23,303 @@
-            {
-                bStrides[i] = descB->mLengths[i - 1] * bStrides[i - 1];
-            }
-#else // HIPTENSOR_DATA_LAYOUT_COL_MAJOR
-            for(int i = modeSize - 2; i >= 0; i--)
+        static constexpr int NumInput  = InDataTypeTuple::Size();
+        static constexpr int NumOutput = OutDataTypeTuple::Size();
+
+        // Argument
+        struct Argument : public BaseArgument
+        {
+             Argument(const std::array<index_t, NumDim> lengths,
+                      const std::array<std::array<index_t, NumDim>, NumInput> inStridesArray,
+                      const std::array<std::array<index_t, NumDim>, NumOutput> outStridesArray,
+                      const void* in_dev_buffers,
+                      void* out_dev_buffers,
+                      ElementwiseOperation elementwise_op,
+                      UnaryOperation unary_op,
+                      Scale scale_op)
+                    : BaseArgument()
+                    , mLengths(lengths)
+                    , mInStrides(inStridesArray)
+                    , mOutStrides(outStridesArray)
+                    , mElementOp(elementwise_op)
+                    , mUnaryOp(unary_op)
+                    , mScaleOp(scale_op)
             {
-                bStrides[i] = descB->mLengths[i + 1] * bStrides[i + 1];
+                mInput  = (mInDataType*)in_dev_buffers;
+                mOutput = (mOutDataType*)out_dev_buffers;
             }
-#endif // HIPTENSOR_DATA_LAYOUT_COL_MAJOR
-            auto  bIndices     = std::vector<int32_t>(modeSize, 0);
-            auto  elementCount = hiptensor::elementsFromLengths(aLens);
-            float alphaValue   = readVal<float>(alpha, typeScalar);
-            for(int elementIndex = 0; elementIndex < elementCount; elementIndex++)
+
+            Argument(Argument const&)            = default;
+            Argument& operator=(Argument const&) = default;
+            ~Argument()                          = default;
+
+            const mInDataType *mInput;
+            mOutDataType      *mOutput;
+
+            std::array<index_t, NumDim>                         mLengths;
+            std::array<std::array<index_t, NumDim>, NumInput>   mInStrides;
+            std::array<std::array<index_t, NumDim>, NumOutput>  mOutStrides;
+
+            ElementwiseOperation mElementOp;
+            UnaryOperation       mUnaryOp;
+            Scale                mScaleOp;
+        };
+
+        // Invoker
+        struct Invoker : public BaseInvoker
+        {
+            using Argument = ReferencePermutation::Argument;
+
+            float Run(const Argument& arg)
             {
-                auto index = elementIndex;
+                int     modeSize        = arg.mLengths.size();
+                auto    elementCount    = hiptensor::elementsFromLengths(std::vector(std::begin(arg.mLengths), std::end(arg.mLengths)));
+
+                std::vector<int> outStrides(std::begin(arg.mOutStrides[0]), std::end(arg.mOutStrides[0]));
+
```

Comment:
Would like a quick comment for description of the following code's intention and calculations

Reply:
Yes its simplified and comments added in the latest commit